rust-connectors/sinks/kafka/src/main.rs

```diff
@@ -18,15 +26,55 @@ async fn main() -> anyhow::Result<()> {
         println!("{}", serde_json::to_string_pretty(&schema).unwrap());
         return Ok(());
     }
-    let opts: KafkaOpt = KafkaOpt::from_args();
-    opts.common.enable_logging();
+    let raw_opts = KafkaOpt::from_args();
+    raw_opts.common.enable_logging();
     info!(
         connector_version = env!("CARGO_PKG_VERSION"),
         git_hash = git_hash_version(),
         "Starting Kafka sink connector",
     );
 
-    opts.execute().await?;
+    // =======================================================
+    // Setup objects and actors based on opts
+    // =======================================================
+
+    // KafkaSinkWorker members need to be inside Arc because it is referenced inside tokio::spawn
+    // That expect things moved inside to be static
+    // An optimization point for this would be to own the runtime (rather than calling tokio::*)
+    // So that runtime::spawn does not need producer to be in Arc
+    let kafka_sink_deps =
+        KafkaSinkDependencies::from_kafka_opt(raw_opts).await? as KafkaSinkDependencies;
+
+    // =======================================================
+    // Streaming sequence starts here
+    // =======================================================
+
+    info!("Starting stream");
+
+    let mut stream = kafka_sink_deps
+        .common_connector_opt
+        .create_consumer_stream()
+        .await?;
+
+    while let Some(Ok(record)) = stream.next().await {
+        // Clone arcs to be sent to async block's scopes
+        let producer = kafka_sink_deps.kafka_producer.clone();
+        let kafka_partition = kafka_sink_deps.kafka_partition.clone();
+        let kafka_topic = kafka_sink_deps.kafka_topic.clone();
+
+        tokio::spawn(async move {
```

Comment:
> It's a little unclear to me what happens when you don't await on any of the spawned tasks.

Without `tokios::spawn`, it would be a stagnant future - not executed by anything. Awaiting a future is essentially polling it until it yields `Poll::Ready`. Moving it using `tokio::spawn` delegates the polling to tokio's default runtime.

> Also, we don't get any errors if we don't await.

There are two futures involved here. 1.) the `async move block` and 2.) `let future = producer.send`.
Future 1 is going to be handled by the built-in tokio runtime.
Future 2 is awaited inside of Future 1, which is in our control. On line 53, the same file, we can catch `res` when it is `Err`. Error handling can be done here as follows:
We can build a batched error pool. Said error pool is processed on the side, concurrent with the main loop. On `Err() = res` the Err value can be then piped into the error pool.

Reply:
There is a fundamental problem here.  Since this spawns new future tasks, it could cause an unlimited number of tasks that could cause OOM or other issues.  Not only that, if Kafka goes down or returns some other error, this would just create an endless number of tasks that just spit out errors.

Reply:
I agree this requires a backpressure mechanism. I'll search for it then. My intuition is to add a limit to the max number of task in a tokio pool. I can't find the API to get that count, so I think I would need to implement that with a pool of JoinHandle and select! inside the main loop to context switch between 1.) the piping of events to kafka and 2.) "garbage-collecting" the finished JoinHandles in the pool

Btw, CMIIW, I assume tokio's built-in would automatically drop the resolved futures regardless the JoinHandle is awaited or not. Would you know if that assumption is correct or not? I'll ask the forum next about it.

Reply:
I don't think it's a good idea to rely on Tokio pool since we might move to another engine in the future. Instead, it would be better to build application-specific code to deal with this.

Reply:
Ok, what about a pool of pairs of JoinHandle and a sum of approximate memory usage of the related future

The additional parameter would be the max memory usage (with caveat that the calculation is an approximate and we describe how we are calculating it and the limitation and pitfall of the parameter)

Reply:
How about using a channel?

Reply:
> How about using a channel?

Sorry, I don't follow. Could you elaborate more on using channel?

Btw, the solution I'm thinking that utilize the approximate memory usage looks like this https://github.com/Kelerchian/fluvio-connectors/tree/290-kafka-sink-batch-with-backpressure

Disclaimer, that branch is not finished yet

Reply:
(I commented several days ago but it was "pending" and did not realize what it meant, probably you couldn't read that one)